### ISSUE TITLE
Python Launcher for Windows

### ISSUE THREAD

**Issue by gschwaer** [sio/Makefile.venv#15]
Hi, I stumbled upon a problem: Recently Windows user tend to use the [Python Launcher for Windows](https://docs.python.org/3.10/using/windows.html#python-launcher-for-windows) to execute python programs. So they use, e.g., `py -3 script.py` instead of `python3 script.py` to launch a script with Python 3.

Unfortunately that also means that the python executable is not be on the path with its *nix/Linux name.

This leads to a problem due to `PY` using the *nix/Linux name
https://github.com/sio/Makefile.venv/blob/bbb679568c2350e013e0e58e8e582f9e4e7091b1/Makefile.venv#L44-L47
but this name is not on the path on windows systems using `py`, so this will fail
https://github.com/sio/Makefile.venv/blob/bbb679568c2350e013e0e58e8e582f9e4e7091b1/Makefile.venv#L90-L93

It is very annoying that Windows continuously creates problems for cross-platform tools (like Makefile.venv). I don't have a good idea how to solve that. Maybe detect if `py.exe` can be found on `PATH` before the OS detection and if so replace 'python' with 'py -' in `PY`? What do you think?

I built this snippet that seems to work:
```Makefile
# Detect Windows py launcher and use it to find the given python version
ifeq ($(OS),Windows_NT)
ifneq (, $(shell where py 2>NUL))
PY := $(subst python,py -,$(PY))
endif
endif
```

Some more infos:
* running `py -3.7 -m venv .venv` results in a proper `.venv/` with `python.exe` and `pip.exe` in `.venv/Scripts/`
* running `py -3.7 -c "import __future__, sys; print(sys.platform)"` prints `win32`
* related to #2

**Comment by sio** [comment]
> They use, e.g., `py -3 script.py` instead of `python3 script.py` to launch a script with Python 3.

That's exactly the reason why PY variable exists. If user needs some magic to invoke Python interpreter they need to be explicit about it.

Does the following work? I think it should.

```shell
$ export PY='py -3'
$ make venv
```

> ifeq ($(OS),Windows_NT)

Sadly it's not that easy: 

- [Cygwin](https://github.com/sio/Makefile.venv/issues/2) does not use "*.exe" suffix but has the same $OS value,
- [FreeBSD](https://github.com/sio/Makefile.venv/commit/8f98229cddce111e6328d2a3b60024103c077766) has no $OS variable at all (though that's easy to work around),
- and on top of that someone might want to invoke native win32 Python from GNU Make running in Cygwin environment...

Basically, I've given up on auto detection of Python command in Windows because there is neither a default way to invoke the interpreter nor a sane algorithm to find one. I'm open to suggestions but they should not break currently working use cases (using $OS breaks both Cygwin and win32 interpreter on Cygwin scenarios)

**Comment by gschwaer** [comment]
> Does the following work? I think it should.

It does under Windows, but not under Linux.

> Sadly it's not that easy:

Hm, I think I didn't explain my idea well. What you describe is indeed a problem if we replace
https://github.com/sio/Makefile.venv/blob/bbb679568c2350e013e0e58e8e582f9e4e7091b1/Makefile.venv#L90-L93
with that snippet I wrote. But that was not my intention. I was proposing to add that snippet before those lines. This way the user's Makefile will use, e.g., `PY=python3` and Makefile.venv will automatically correct that to `PY=py -3` if it detects the Python windows launcher.

> Basically, I've given up on auto detection of Python command in Windows because there is neither a default way to invoke the interpreter nor a sane algorithm to find one.

This is exactly the Problem that the Python people paced as well and is why they developed `py` that does that job for you. What I proposed is just: If `py` is existing on the system: use it.

> I'm open to suggestions but they should not break currently working use cases (using $OS breaks both Cygwin and win32 interpreter on Cygwin scenarios)

I disagree, but I think your reasoning here is based on the assumption of replacing the quoted lines above. In case you are on any platform that does not have `ifeq ($(OS),Windows_NT)`, the behavior does not change. Else if you have not installed `py`, the behavior will not change. (The check for `$(OS) == Windows_NT` is just for `$(shell where py 2>NUL)` to work, because Windows has `where` not `which`.)
The only case that changes, is if you are on Windows and have `py` installed. In this case Makefile.venv will use `py` to execute the (in `PY`) requested python version.

An even more restrictive version would also check if the python executable in `PY` was not found, like so:
```Makefile
# Detect Windows py launcher and use it to find the given python version
ifeq ($(OS),Windows_NT)
ifeq (, $(shell where $(PY) 2>NUL))
ifneq (, $(shell where py 2>NUL))
PY := $(subst python,py -,$(PY))
endif
endif
endif
```
(Untested) (It will only use `py` if you are 1. on windows, 2. the specified version is not on PATH, and 3. `py` is on PATH. It could be made even more robust by checking if `PY` contains no slashes and starts with `python`, ...)

**Comment by sio** [comment]
Thanks for explaining this further. Now I see that it's doable :-)

I'm busy with other things at the moment, will tinker with this next week.

**Comment by gschwaer** [comment]
Sounds good, thank you for being open minded about that.
This change would really make `Makefile.venv` to the cross-plattform tool that fits all our needs. :ok_hand:

**Comment by sio** [comment]
I've added [autodetection of Python interpreter](https://github.com/sio/Makefile.venv/commit/578996511d1ad784aaefbea9aa8e38f91d117b85) to a feature branch this week. [Tests](https://github.com/sio/Makefile.venv/commit/679b14ccfcb29656c718facdf788b3480b95e126) are not green yet and still need some ironing out in regards to Windows quirks.

**Comment by gschwaer** [comment]
I like the idea, but the current implementation actually doesn't solve the issue I had. You now try to find the python executable but the user cannot specify its version.

The thing is that I set `PY=python3.7`, but the system doesn't have that on PATH. Only `py` can find the location of python3.7. So if I set `PY` to `python3.7` the whole detection is ignored.

Hm, maybe it would work if you instead let the user makefile specify the python version, like `PY_VERSION` and then search for that in `.venv/[bin/Script]/python$PY_VERSION`, then test PATH with `python$PY_VERSION`, then test `py -$PY_VERSION`. What do you think?

**Comment by sio** [comment]
I'm not sure I'm following.

First of all, you do understand that PY only matters for initial venv creation, right? All subsequent calls will be executed by the interpreter provided by venv, there is no need to dive into specifics (`.venv/[bin/Script]/python$PY_VERSION`) because venv should contain only one interpreter version.

> But the user cannot specify its version. The thing is that I set `PY=python3.7`, but the system doesn't have that on PATH.

You need to explain which hat are you wearing here? Are you a Makefile developer or are you an end user who just executes a Makefile given to them?

- If you're end user - you should know where in the system your preferred interpreter is located. Put the full path to that interpreter into PY environment variable before running `make`. Using `PY="py -3.7"` will save you some keystrokes but is equivalent otherwise.
- If you're a Makefile developer - why would you want to require `python3.7` specifically instead of "the newest installed Python 3" `py -3` currently gives you? How could Python 3.8 or 3.10 be detrimental in your case?

**Comment by gschwaer** [comment]
> First of all, you do understand that PY only matters for initial venv creation, right?

You are right, I got confused there, probably because I saw this in your changes
https://github.com/sio/Makefile.venv/blob/578996511d1ad784aaefbea9aa8e38f91d117b85/Makefile.venv#L95-L109
If `PY` is only used for venv creation, why those lines? Both `$(VENVDIR)/bin/python` and `$(VENVDIR)/Script/python` will not exist, right?

Actually `PY` is also used for OS detection on every make command after the initialization of venv. That is why the following produces an error message on master (the venv already exists):
```bash
make docs
# normal output
make docs PY=not_existing
/bin/sh: 1: not_existing: not found
# normal output
```
On windows I assume it would even fail, because `VENV` and `EXE` are not set correctly.
https://github.com/sio/Makefile.venv/blob/578996511d1ad784aaefbea9aa8e38f91d117b85/Makefile.venv#L140-L146

>> But the user cannot specify its version. The thing is that I set PY=python3.7, but the system doesn't have that on PATH.
>
> You need to explain which hat are you wearing here? Are you a Makefile developer or are you an end user who just executes a Makefile given to them?

Sorry, with "user" I meant the Makefile developer. What I wanted to say in the sentence above: As a Makefile developer using Makefile.venv with your changes, I can either give a python executable in `PY` (e.g., `PY=python3.7`) but that has to be on PATH on Linux and Windows. Or I don't set `PY`, in which case I get some (for the Makefile developer) unknown python version by the autodetection.

> If you're a Makefile developer - why would you want to require python3.7 specifically instead of "the newest installed Python 3" py -3 currently gives you? How could Python 3.8 or 3.10 be detrimental in your case?

Because we are developing against that version because it is the oldest version we guarantee to support. If we all develop on a newer version and by accident add some code that doesn't work on 3.7, we will get flooded with bug reports. So we want e.g., all our tests to run on 3.7 with `make test`.

> I'm not sure I'm following.

I was explaining poorly, sorry. Here is the situation:

We develop in a team (some on Linux, some on Windows) and are restricted to a specific python version (3.7). I want that version to be in the Makefile, so nobody accidentally uses a wrong version. After the venv is created with the correct python version all make targets run to satisfaction as is.

But for the creation of the venv, I need to select the correct python version. Well, Linux has a nice `PATH` that is generally accepted and used to detect executables, but Windows doesn't. Or at least enough people think so, which is why they wrote their own detection tools (`py` launcher tool) and don't use the Windows PATH.

Now if I want all make targets to use Python 3.7, on Linux it is sufficient to set `PY=python3.7` in our project Makefile, but on Windows, it cannot be found on PATH. So my Windows colleagues will have to pass `PY="py -3.7"` on every make command to override it, which I will have to document and maybe someone accidentally use a wrong version, ...

So my idea was that when Makefile.venv receives a python executable name in `PY`, but cannot find it on path, it will try to use an alternate means of python executable detection, that is the `py` launcher tool. But it should preserve the version that was provided in `PY`. Otherwise the Windows colleagues will use a different version than the Linux colleagues without knowing.

Well that became a long message :sweat_smile: I hope I made my thoughts more clear though. Thank you for your patience and efforts on this!

**Comment by sio** [comment]
> If PY is only used for venv creation, why those lines?
> <...>
> Actually PY is also used for OS detection on every make command after the initialization of venv. That is why the following produces an error message on master

I took an opportunity to squash that error while adding the interpreter detection logic :) Now if venv already exists, Windows autodetection will never fail.

> Because we are developing against that version because it is the oldest version we guarantee to support. If we all develop on a newer version and by accident add some code that doesn't work on 3.7, we will get flooded with bug reports.

Hardcoding Python version into your Makefile is a poor practise. PY was never meant for that. You should better move that logic into your CI environment variables, and specify Python interpreter (via PY) there.

As for local developer's workstations: you can enforce any rules you want, there is no need to change upstream Makefile.venv.

You can error out if you don't like the Python version:

```make
ifeq (False,$(shell $(PY) -c "import sys; print(sys.version_info[:2] == (3,7))"))
$(error Must use Python 3.7! Provide path/command for python3.7 via PY environment variable)
endif
```

You can supply your own autodetection logic prior to including Makefile.venv:

```make
PY_VERSION=3.7
ifeq (ok,$(shell python$(PY_VERSION) -c "print('ok')"))
PY=python$(PY_VERSION)
else
PY=py -$(PY_VERSION)
endif
```

But I wouldn't do any of that unless you're sure that your Makefile will never be executed by someone outside of your developer team. And even then, what if some obscure bug affects only Python 3.8 and is not present on 3.7? How would your developer debug that issue if you'd hardcode Python 3.7?

I think minimal supported interpreter version should be enforced on CI runner, not on individual developer workstations. If at some point one of your developers is hitting 3.7 incompatibilities often enough, nothing is stopping them from doing `export PY=python3.7` for a session.

**Comment by gschwaer** [comment]
Ok, I see what you mean. My thought was, that CI can overwrite `PY` when testing, but the developers should not have to set it manually, when 3.7 is our main python target either way. You are right, I can do that in the project Makefile prior to including Makefile.venv. I was just hoping I can do that in a platform agnostic way, and Makefile.venv will handle the platform specific part, but with `PY` I will have to pass the interpreter which is platform specific (`python`/`py`).
Thanks for your time and all the efforts you put in this project. It is greatly appreciated :+1:

**Comment by sio** [comment]
I've merged `py` support into master and published a [new version](https://github.com/sio/Makefile.venv/tree/v2021.12.16).

Python version enforcing is definitely out of scope for Makefile.venv, such stuff is better left out for individual purpose-built Makefiles.
